### ISSUE TITLE
Callbacks for every epoch 

### ISSUE THREAD

**Issue by nithila-yalamanchi** [autogluon/autogluon#392]
Hey guys, so I've been using autogluon for a while now but I unable to figure out a way to have a Hook/Callback which is called after every epoch and after every model is trained. So, I'd highly appreciate if you guys could point me to the appropriate method if it is already implemented else please let me know how to add my own callbacks function.

**Comment by mseeger** [comment]
Hello, your training evaluation function calls the reporter after each epoch, passing on data, which has to include the validation metric (such as accuracy -- careful, the metric must be such that larger is better). Have a look at an example, and note the reporter is passed as argument to the evaluation function.

You can then look into the code for schedulers (e.g. FIFOScheduler and HyperbandScheduler), and check the _run_reporter methods. This is where this data is coming back. It then influences how scheduler (and searcher) operate in the future.

This information should help you making progress. Of course, you need to somehow stay within the space of what AG is doing, unless you totally want to rewrite everything.

**Comment by nithila-yalamanchi** [comment]
Hey, thanks for the quick response. I’ve tried this method and these seem to work if we have our own training function, but for some components in my project, I’m using TabularPrediction.fit() and I need to have the callbacks mechanism for the same on tabular data. Though I’m able to get fit_summary (after training with final accuracy values for each model but not with intermediate values), I need to have realtime metrics after each epoch(if possible for each model fitted). Something similar to the code snippet below.

Def myCallbacksFn(eval_result):
print(“Called the callback”)
<My Code which uses eval_result for logging, GUI plotting, etc>
train_data = task.Dataset(file_path='<dataset_path>')
Predictor = TabularPrediction.fit(train_data = train_data, callbacks=[myCallbacksFn,])

So, myCallbacksFn will be called after each epoch in each model. Is this possible? I’ve studied the TabularPredictions code recursively and found that Catboost, tabular_nn, etc code is not equipped to call any certain function after an iteration/epoch.

**Comment by Innixma** [comment]
Hi Nithila,

Thanks for the question and feedback! Tabular currently doesn't support user-specified callback functions. I am unsure if this is the case for ImageClassification as well or exclusive to Tabular.

Adding custom callback support gets a bit tricky since not all models have a concept of an 'iteration', and in particular, I don't believe CatBoost has the ability to have a callback function, so adding it for AutoGluon is not possible until the CatBoost developers implement such functionality (Which I would be very happy to utilize). LightGBM does have callback support, which we use to track memory usage and perform early stopping. TabularNN would also be able to have callback support implemented. RF could have a callback, but it would dramatically slow down training and not offer much value. Others such as KNN cannot have a callback because they don't have iterations.

We would have to identify what are the expected inputs to your custom callback function, and what influence you want it to be able to have over the training process. (Purely logging or functionality such as determining early stopping). Would callbacks be specified for a specific model, or across all models?

This logic would likely live in scheduler. @zhanghang1989 , what is the required work to enable custom callback functionality to scheduler, such that it is called after every iteration? How would this work in distributed settings?

**Comment by jwmueller** [comment]
Like Nick said, TabularPrediction uses many models by default, many of which are not trained in "epochs".  If you're ok to solely use the neural network model in TabularPrediction (this will result in lower accuracy), then you can record per-epoch callbacks with these steps:

1) Install autogluon from source (and install mxboard):
```
pip install --upgrade mxnet mxboard
git clone https://github.com/awslabs/autogluon
cd autogluon && python setup.py develop
```

2) Go to this line of code: https://github.com/awslabs/autogluon/blob/master/autogluon/utils/tabular/ml/models/tabular_nn/tabular_nn_model.py#L339
and modify it so that your custom metric is calculated and also added to the `summary_writer`.  In order to obtain predictions from this epoch to use in your custom metric calculation, you can call `pred = self.predict_proba(x)`, with x = `train_dataset` or `test_dataset`.

3) Run your modified version of autogluon like this:
```
from autogluon import TabularPrediction as task
predictor = task.fit(..., hyperparameters={'NN':{'num_epochs: 500}}, verbosity=4)
```

The callback values should be available in real-time through the mxboard interface: https://github.com/awslabs/mxboard

**Comment by arunmk** [comment]
@Innixma @jwmueller is it possible for Tabular to have pre and post callback functions for each model fit() run, so that we may gather data even while the entire set of models is running.

**Comment by Innixma** [comment]
Hi Arun, please see my update to https://github.com/awslabs/autogluon/issues/492
